File: problems/leetcode/0700/787/787.py
# 787. Cheapest Flights Within K Stops
# https://leetcode.com/problems/cheapest-flights-within-k-stops/description/

from collections import defaultdict, deque
from typing import *

# BFS 방식 탐색

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        flightMap = defaultdict(defaultdict[int])
        for f, t, p in flights:
            flightMap[f][t] = p

        cheapest = dict()
        for i in range(n):
            cheapest[i] = 0 if i == src else float("inf")

        q = deque() # (nodeNum, currentPrice, currentStop)
        q.append((src, 0, 0))

        minPrice = float("inf")
        while len(q) > 0:
            nodeNum, currentPrice, currentStop = q.popleft()

            if nodeNum == dst:
                minPrice = min(minPrice, currentPrice)
                continue

            for nextNode, price in flightMap[nodeNum].items():
                if nextNode == dst or (currentStop < k and currentPrice + price < cheapest[nextNode]):
                    cheapest[nextNode] = currentPrice + price
                    q.append((nextNode, currentPrice + price, currentStop + 1))

        return -1 if minPrice == float("inf") else minPrice

File: problems/leetcode/0700/787/787.cpp
// 787. Cheapest Flights Within K Stops
// https://leetcode.com/problems/cheapest-flights-within-k-stops/

using namespace std;

#include <map>
#include <queue>
#include <vector>

typedef pair<int, pair<int, int>> piii;

class Solution {
public:
  int findCheapestPrice(int n, vector<vector<int>> &flights, int src, int dst,
                        int k) {
    map<int, vector<pair<int, int>>> flightPrices;
    for (auto &flight : flights) {
      int from = flight[0], to = flight[1], price = flight[2];
      flightPrices[from].push_back({to, price});
    }

    int currentStops[100];
    fill_n(currentStops, n, INT_MAX);

    priority_queue<piii, vector<piii>, greater<piii>>
        pq; // { price, { city, stops } }
    pq.push({0, {src, 0}});

    while (!pq.empty()) {
      int price = pq.top().first, city = pq.top().second.first,
          stops = pq.top().second.second;
      pq.pop();

      if (city == dst) {
        return price;
      }

      if (currentStops[city] < stops) {
        continue;
      }

      currentStops[city] = stops;

      if (stops > k) {
        continue;
      }

      for (auto &flight : flightPrices[city]) {
        pq.push({price + flight.second, {flight.first, stops + 1}});
      }
    }

    return -1;
  }
};
